signature-specification.md

```diff
@@ -0,0 +1,519 @@
+
+Pros:
+
+- Simple, lesser probability of user shooting themselves in the foot.
+- Signing libraries available in Go: [lib/paseto](https://pkg.go.dev/github.com/shuLhan/share/lib/paseto?utm_source=godoc), [o1egl/paseto](https://github.com/o1egl/paseto).
+
+Cons:
+
+- No published RFC, although there is one in [draft](https://datatracker.ietf.org/doc/html/draft-paragon-paseto-rfc-00).
+- Doesn't provides algorithm agility instead uses predefined algorithm (for recommended version v2, signing algorithm is `Ed25519`) which might not be supported by smart cards, HSMs and cloud based Key Management Systems. Also given that v2 only support `Ed25519` signing algorithm, which is not FIPS 140-2 approved thus system using it cannot be FIPS 140-2 compliant.
+- No native support for referencing or inlining of certificate(s), keyId and timestamp (TSA). These are supported through a signed footer which can contain any structure.
+- No native support for unsigned attributes and TSA signature.
+- Doesn't support multiple signature in one signature envelope.
+
+<details><summary>Example (Click to show details)</summary>
+
+`payload`:
+
+```json
+{ 
+    "subjectManifest": {...},
+    "signedAttrs": {
+        "reserved": {
+            "iat": "2021-06-22T16:02:40.3375379Z",
+            "exp": "2021-09-20T16:02:40.3375379Z",
+            ...
+        },
+        "custom" : {...}
+    }
+}
+```
+
+`footer`:
+
+```json
+{
+    "x5c": [
+        "<Base64(DER(leafCert))>",
+        "<Base64(DER(intermediateCACert))>",
+        "<Base64(DER(rootCert))>"
+    ],
+    "timestamp": "<Base64(TimeStampToken)>"
+}
+```
+
+`Signature Envelope`:
+
+```
+<version>.<purpose>.Base64Url(<payload><Sign(PAE(<version><purpose><payload>))>).Base64Url(<footer>)
+```
+
+- `version`: `v2`, v2 is recommended version.
+- `Purpose`: `public`, public is used for signing.
+- `PAE`: [Pre-Authentication Encoding](https://tools.ietf.org/id/draft-paragon-paseto-rfc-00.html#rfc.section.2.2.1)
+
+</details>
+
+
+### Red Hat's Simple Signing - [Simple Signing](https://www.redhat.com/en/blog/container-image-signing)
+
+It is used by Red Hat for signing container image. Simple signing is *not a signature format* instead defines the payload format which is then signed using PGP to create detached signature. The payload format is adapted by atomic and later by [cosign](https://github.com/sigstore/cosign#payload-formats).
+
+<details><summary>Example (Click to show details)</summary>
+
+`payload`:
+
+```json
+{
+    "critical": {
+        "identity": {
+            "docker-reference": "testing/manifest"
+        },
+        "image": {
+            "Docker-manifest-digest": "sha256:20be...fe55"
+        },
+        "type": "atomic container signature"
+    },
+    "optional": {
+        "creator": "atomic",
+        "timestamp": 1458239713
+    }
+}
+```
+
+</details>
+
+
+## Recommendation
+
+### Signature format
+
+Recommendation is to **use JWS JSON Serialization** because of following reasons:

```

Comment:
We should clarify this is the recommendation of the Microsoft and AWS crypto boards, and not that of the general Notary project community, unless we get consensus among the community.